### PR TITLE
Remove unnecessary compatibilty check for MaterializedMySQL depends on MySQL server version

### LINKED ISSUE #63268: MaterializedMySQL won't start with MySQL8.4
**Describe what's wrong**

Recently released MySQL 8.4
remove `--default-authentication-plugin=mysql_native_password`
and replace it to
`--mysql-native-password=ON`

> A link to reproducer 

https://gist.github.com/Slach/1db8f5ec1df586f5f6427b0188977fdd 

**Does it reproduce on the most recent release?**
Yes, it reproduce on head and latest docker images (24.4,24.5)

**How to reproduce**
```
git clone https://gist.github.com/Slach/1db8f5ec1df586f5f6427b0188977fdd ./mysql84_materialized_mysql
cd ./mysql84_materialized_mysql
bash -xe ./reproduce.sh
docker-compose exec clickhouse tail -n 100 /var/log/clickhouse-server/clickhouse-server.err.log
```

**Expected behavior**
Successful start `MaterializedMySQL`

**Error message and/or stacktrace**

```
2024.05.02 10:06:28.851189 [ 48 ] {2eb2c123-b147-4e67-a30d-7279d059e23a} <Error> TCPHandler: Code: 695. DB::Exception: Load job 'startup MaterializedMySQL database mysql' failed: Code: 537. DB::Exception: Illegal MySQL variables, the MaterializedMySQL engine requires default_authentication_plugin='mysql_native_password'. (ILLEGAL_MYSQL_VARIABLE), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c9b62fb
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000078128ec
2. DB::Exception::Exception<String&>(int, FormatStringHelperImpl<std::type_identity<String&>::type>, String&) @ 0x000000000782d3eb
3. DB::checkMySQLVariables(mysqlxx::Pool::Entry const&, DB::Settings const&) @ 0x000000001017d8c1
4. DB::MaterializedMySQLSyncThread::assertMySQLAvailable() @ 0x000000001017c0e9
5. void std::__function::__policy_invoker<void (DB::AsyncLoader&, std::shared_ptr<DB::LoadJob> const&)>::__call_impl<std::__function::__default_alloc_func<DB::DatabaseMaterializedMySQL::startupDatabaseAsync(DB::AsyncLoader&, std::unordered_set<std::shared_ptr<DB::LoadJob>, std::hash<std::shared_ptr<DB::LoadJob>>, std::equal_to<std::shared_ptr<DB::LoadJob>>, std::allocator<std::shared_ptr<DB::LoadJob>>>, DB::LoadingStrictnessLevel)::$_0, void (DB::AsyncLoader&, std::shared_ptr<DB::LoadJob> const&)>>(std::__function::__policy_storage const*, DB::AsyncLoader&, std::shared_ptr<DB::LoadJob> const&) @ 0x0000000010140ca3
6. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::AsyncLoader::spawn(DB::AsyncLoader::Pool&, std::unique_lock<std::mutex>&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000cbb1106
7. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000ca6d919
8. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000ca7168a
9. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000ca7048d
```

**Additional context**

Will try to prepare PR soon

### PR DESCRIPTION
**Pull request by Slach**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove unnecessary compatibilty check for MaterializedMySQL depends on MySQL server version, fix https://github.com/ClickHouse/ClickHouse/issues/63268

### CODE REVIEW

File: src/Databases/MySQL/MaterializedMySQLSyncThread.cpp [hunk 1]
```diff
@@ -165,8 +161,41 @@ static void checkMySQLVariables(const mysqlxx::Pool::Entry & connection, const S
 
         for (size_t index = 0; index < variables_block.rows(); ++index)
         {
-            const auto & error_message_it = variables_error_message.find(variable_name_column->getDataAt(index).toString());
+            const String variable_name = variable_name_column->getDataAt(index).toString();
+            if variable_name == "version" {
+                version = variable_value_column->getDataAt(index).toString();
```
Comment (Enmk) [inline src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:166]: Maybe add a `break` here? Or is there a good reason to check all rows?

File: src/Databases/MySQL/MaterializedMySQLSyncThread.cpp [hunk 2]
```diff
@@ -165,8 +161,41 @@ static void checkMySQLVariables(const mysqlxx::Pool::Entry & connection, const S
 
         for (size_t index = 0; index < variables_block.rows(); ++index)
         {
-            const auto & error_message_it = variables_error_message.find(variable_name_column->getDataAt(index).toString());
+            const String variable_name = variable_name_column->getDataAt(index).toString();
+            if variable_name == "version" {
```
Comment (Enmk) [inline src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:165]:
```suggestion
            if (variable_name == "version")
            {
```

File: src/Databases/MySQL/MaterializedMySQLSyncThread.cpp [hunk 3]
```diff
@@ -165,8 +161,41 @@ static void checkMySQLVariables(const mysqlxx::Pool::Entry & connection, const S
...
+        }
+
+        if (version >= "8.4.0")
+        {
+            variables_error_message = {
+                {"log_bin", "ON"},
+                {"binlog_format", "ROW"},
+                {"binlog_row_image", "FULL"},
+            };
+        } else if (versin >= "8.3.0") {
```
Comment (Enmk) [inline src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:178]:
```suggestion
        }
        else if (version >= "8.3.0")
        {
```

File: src/Databases/MySQL/MaterializedMySQLSyncThread.cpp [hunk 4]
```diff
@@ -165,8 +161,41 @@ static void checkMySQLVariables(const mysqlxx::Pool::Entry & connection, const S
...
+            variables_error_message = {
+                {"log_bin", "ON"},
+                {"binlog_format", "ROW"},
+                {"binlog_row_image", "FULL"},
+                {"default_authentication_plugin", "mysql_native_password"},
+            };
+        } else {
```
Comment (Enmk) [inline src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:185]:
```suggestion
        }
        else
        {
```

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 1f1eb12fb7d597d8962e1ae0c99f7d3a04726bab with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63269/1f1eb12fb7d597d8962e1ae0c99f7d3a04726bab/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Databases/MySQL/MaterializedMySQLSyncThread.cpp [hunk 5]
```diff
@@ -165,8 +162,47 @@ static void checkMySQLVariables(const mysqlxx::Pool::Entry & connection, const S
...
+        else if (versin >= "8.3.0")
+        {
+            variables_error_message = {
+                {"log_bin", "ON"},
+                {"binlog_format", "ROW"},
+                {"binlog_row_image", "FULL"},
+                {"default_authentication_plugin", "mysql_native_password"},
+            };
+        }
+        else
```
Comment (Enmk) [inline src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:190]: Due to peculiarities of string comparison, any version starting from "10.0.0" would end up at this part, not in `>= "8.4.0"`

Comment (Slach) [reply src/Databases/MySQL/MaterializedMySQLSyncThread.cpp:191]: fixed https://github.com/ClickHouse/ClickHouse/pull/63269/commits/834be6de43e8b05a0a7e46639025943f0a4199a9

**Comment by alexey-milovidov** [comment]
@Slach @Enmk tests have failed.

**Comment by Slach** [comment]
will try to add changes for integration tests

**Comment by alexey-milovidov** [comment]
It's unclear how to proceed. Tests have failed, and there has been no activity since May.

**Pull request closed without merging**